### PR TITLE
fix(electric): Fix column name quoting in one of the trigger functions

### PR DESCRIPTION
**Pull request by alco**
Fixes VAX-1062.

I had to refactor the way the affected SQL function is defined in order for it to work both when running Electric for the first time and when upgrading from an earlier version. I recommend reviewing this PR one commit at a time to follow the progression of function code repackaging.

### CODE REVIEW

**Comment by alco** [comment]
Hey @icehaunter. Could you take a look at the new E2E test and the logs for `electric`? While writing the test code, I had an iteration that was failing without the bug fix and passing after applying the bug fix. But I have made some changes to it since to verify that a value in the column named `"userId"` can be updated and now the test is failing again.

I probably have wrong operations in my `Satellite.TestWsClient.send_data()` call. The test code expects the row with ID=1 to have its userId updated from `NULL` to `"test_user"`:
```
[shell pg_1]
    [invoke wait-for "SELECT * FROM public.\"Foo\";" "test_user" 10 ${psql}]

    !SELECT * FROM public."Foo" ORDER BY id;
    ??1 | test_user
    ??2 | pg_user
    ??3 | satellite_user
```

but it never updates. Here's the latest result of the above query from the `pg_1` log:
```
electric=# SELECT * FROM public."Foo";
 id |     userId     
----+----------------
 2  | pg_user
 3  | satellite_user
 1  | <NULL>
(3 rows)
```

Could my "fix" to the SQL function break the way operations are applied in Postgres? Here's what I find weird in the `electric` logs:
* The log entry `Will send 3 to subscriber` (coming from `SlotServer`) shows both `NewRecord{record: %{"id" => "3", "userId" => "satellite_user"}}` and `UpdatedRecord{record: %{"id" => "1", "userId" => "test_user"}}` inside the same transaction. This looks OK to me, both ops are sent by Satellite in the same transaction.
* Postgres executes the transaction and then replays the rows back to Electric over the logical replication stream. I've added a bunch of `IO.inspect()` calls in `LogicalReplicationProducer` to inspect the tuples. The Update message received from Postgres now contains `nil` for `userId` where I was expecting to see `"test_user"`:

      tests-electric_1-1  | update_msg: %Electric.Postgres.LogicalReplication.Messages.Update{
      tests-electric_1-1  |   relation_id: 16650,
      tests-electric_1-1  |   changed_key_tuple_data: nil,
      tests-electric_1-1  |   old_tuple_data: ["1", nil],
      tests-electric_1-1  |   tuple_data: ["1", nil]
      tests-electric_1-1  | }
      tests-electric_1-1  | old_data: %{"id" => "1", "userId" => nil}
      tests-electric_1-1  | data: %{"id" => "1", "userId" => nil}

* In the shadow table update (in the logs it's just below the main row update) I see that the `"test_user"` value is in the "old shadow row" whereas the "new shadow row" has `nil`:

      tests-electric_1-1  | old_data: %{
      tests-electric_1-1  |   "__reordered_userId" => "test_user",
      tests-electric_1-1  |   "_currently_reordering" => "t",
      tests-electric_1-1  |   "_is_a_delete_operation" => "f",
      tests-electric_1-1  |   "_last_modified" => "746",
      tests-electric_1-1  |   "_modified_columns_bit_mask" => nil,
      tests-electric_1-1  |   "_observed_tags" => nil,
      tests-electric_1-1  |   "_resolved" => "f",
      tests-electric_1-1  |   "_tag" => "(\"2023-09-22 13:57:19.477+00\",)",
      tests-electric_1-1  |   "_tag_userId" => "(\"2023-09-22 13:57:19.477+00\",)",
      tests-electric_1-1  |   "_tags" => "{\"(\\\"2023-09-22 13:57:19.477+00\\\",)\"}",
      tests-electric_1-1  |   "id" => "1"
      tests-electric_1-1  | }
      tests-electric_1-1  | data: %{
      tests-electric_1-1  |   "__reordered_userId" => nil,
      tests-electric_1-1  |   "_currently_reordering" => nil,
      tests-electric_1-1  |   "_is_a_delete_operation" => "f",
      tests-electric_1-1  |   "_last_modified" => "747",
      tests-electric_1-1  |   "_modified_columns_bit_mask" => "{f}",
      tests-electric_1-1  |   "_observed_tags" => nil,
      tests-electric_1-1  |   "_resolved" => "f",
      tests-electric_1-1  |   "_tag" => "(\"2023-09-22 13:57:21.611+00\",client_1_1)",
      tests-electric_1-1  |   "_tag_userId" => "(\"2023-09-22 13:57:19.477+00\",)",
      tests-electric_1-1  |   "_tags" => "{\"(\\\"2023-09-22 13:57:19.477+00\\\",)\",\"(\\\"2023-09-22 13:57:21.611+00\\\",client_1_1)\"}",
      tests-electric_1-1  |   "id" => "1"
      tests-electric_1-1  | }

Any suggestions on whether my test code is wrong or the change I made in the SQL function definition has broken one of the conflict resolution triggers?

**Comment by alco** [comment]
@icehaunter Nevermind. I've resolved the issues described above myself.

@magnetised Please review when you get a chance.

File: components/electric/lib/electric/postgres/extension/functions.ex [hunk 1]
```diff
@@ -1,26 +1,68 @@
...
     end
   end
+
+  @doc """
+  Look up the SQL code for a function by its canonical name (basename without extension).
+
+  We catalog all function definitions as files inside the `functions/` subdirectory. A single file usually contains a
+  single function definition but may have more than one if they are all meant to be evaluated as a unit.
+  """
+  @spec by_name(name) :: sql
```
Comment (magnetised) [inline components/electric/lib/electric/postgres/extension/functions.ex:57]: Again, I made very similar changes to the function definition stuff -- removing the custom args etc. to make the process of adding new functions much simpler. Guess you got there first!

File: components/electric/lib/electric/postgres/extension/migrations/20230605141256_electrify_function.ex [hunk 2]
```diff
@@ -52,6 +52,9 @@ defmodule Electric.Postgres.Extension.Migrations.Migration_20230605141256_Electr
       );
       """,
       Extension.add_table_to_publication_sql(electrified_tracking_table),
+      # This function definition is included here because it is referenced in the definition of the electrify() function
+      # below it.
+      Extension.Functions.by_name(:validate_table_column_types),
```
Comment (magnetised) [inline components/electric/lib/electric/postgres/extension/migrations/20230605141256_electrify_function.ex:57]: ooh. interesting.

**Review by magnetised — APPROVED** [review]
considering the work you've just handed me re rebasing kinda wish you'd just stayed in your lane and just fixed the bug! oh well. all good stuff. the perils of long-running feature branches I guess.

**Pull request merged**